keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+    // is inclusive, i.e. it includes the default IP address stored in the
+    // "PodIP" field, and this default IP address must be recorded in the
+    // 0th entry (PodIPs[0]) of the slice. The list is empty if no IPs have
+    // been allocated yet.
+    PodIPs []PodIPInfo `json:"podIPs,omitempty" protobuf:"bytes,6,opt,name=podIPs"`
+```
+
+##### Default Pod IP Selection
+Older servers and clients that were built before the introduction of full dual stack will only be aware of and make use of the original, singular PodIP field above. It is therefore considered to be the default IP address for the pod. When the PodIP and PodIPs fields are populated, the PodIPs[0] field must match the (default) PodIP entry. If a pod has both IPv4 and IPv6 addresses allocated, then the IP address chosen as the default IP address will match the IP family of the cluster's configured service CIDR. For example, if the service CIDR is IPv4, then the IPv4 address will be used as the default address.
+
+#### PodStatus Internal Representation
+The PodStatus internal representation will be modified to use a slice of PodIPInfo structs rather than a singular IP ("PodIP"):
+```
+    // IP address information. Each entry includes:
+    //    IP: An IP address allocated to the pod. Routable at least within
+    //        the cluster.
+    //    Properties: Arbitrary metadata associated with the allocated IP.
+    // Empty if no IPs have been allocated yet.
+    type PodIPInfo struct {
+        IP string
+        Properties map[string]string
+    }
+
+    // IP addresses allocated to the pod with associated metadata.
+    PodIPs []PodIPInfo `json:"podIPs,omitempty" protobuf:"bytes,6,opt,name=podIPs"`
+```
+This internal representation should eventually become part of a versioned API (after a period of deprecation for the singular "PodIP" field).
+
+#### Maintaining Compatible Interworking between Old and New Clients
+Any Kubernetes API change needs to consider consistent interworking between a possible mix of clients that are running old vs. new versions of the API. In this particular case, however, there is only ever one writer of the PodStatus object, and it is the API server itself. Therefore, the API server does not have an absolute requirement to implement any safeguards and/or fixups between the singular PodIP and the plural PodIPs fields as described in the guidelines for pluralizing singular API fields that is included in the [Kubernetes API change quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md).
+
+However, as a defensive coding measure and for future-proofing, the following API version translation logic will be implemented for the PodIP/PodIPs fields:
+
+##### V1 to Core (Internal) Conversion
+- If only V1 PodIP is provided:
+  - Copy V1 PodIP to core PodIPs[0]
+- Else if only V1 PodIPs[] is provided: # Undetermined as to whether this can actually happen (@thockin)
```

Comment:
If the versioned defaulting works, it should not be possible, but good to be safe about it and spec it